File: dataset_preprocess.py
import os
dataset_files=os.listdir("ddos_dataset")
dataset_files.sort()
unprocessed_traffic=open("ddos_dataset/unprocessed_traffic.txt",'w')
for file in dataset_files:
    read_file=open("ddos_dataset/"+file,'r')
    read_str=read_file.readline()
    while read_str:
        timestamp,src,dst=read_str[:-1].split(" ")
        if len(timestamp)!=20 or len(timestamp.split('.'))!=2 or len(timestamp.split('.')[0])!=10:
            continue
        unprocessed_traffic.write(read_str)
        read_str=read_file.readline()
    read_file.close()
unprocessed_traffic.close()
unprocessed_traffic=open("ddos_dataset/unprocessed_traffic.txt",'r')
processed_traffic=open("ddos_dataset/processed_traffic.txt",'w')
readstr=unprocessed_traffic.readline()
start_time=float(readstr[:-1].split(" ")[0])
sliding_window=[]
current_window_idx=0
end_time=0
f=1
while readstr:
    timestamp=float(readstr[:-1].split(" ")[0])
    if timestamp-start_time>60:
        end_time=timestamp
        break
    sliding_window.append(readstr)
    readstr=unprocessed_traffic.readline()
while 1:
    for record in sliding_window:
        timestamp,src,dst=record[:-1].split(" ")
        dst+="_"+str(current_window_idx)+'\n'
        processed_traffic.write(" ".join([timestamp,src,dst]))
    end_time+=1
    start_time +=1
    current_window_idx+=1
    if len(readstr)==0:
        break
    while float(sliding_window[0][:-1].split(" ")[0])<start_time:
        sliding_window.pop(0)
    while readstr:
        timestamp = float(readstr[:-1].split(" ")[0])
        if timestamp>end_time:
            break
        sliding_window.append(readstr)
        readstr=unprocessed_traffic.readline()
processed_traffic.close()
unprocessed_traffic.close()

File: dataset_preprocess.cpp
#include <cstdint>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>
#include <algorithm>
#include <deque>
using namespace std;
int main(){
    ofstream out;
    ifstream inf;
    string caida_dir_path = "ddos_dataset";
    out.open("processed/unprocessed_traffic.txt", ios::out);
    for (const auto& entry : filesystem::directory_iterator(caida_dir_path)) 
    {
        inf.open(entry.path(), ios::in);
        cout << "Reading " << entry.path() << endl;
        string timestamp,src_ip, dst_ip;
        while (inf >> timestamp && inf >> src_ip && inf >> dst_ip) {
            if(timestamp.size()!=20||timestamp.find_first_of(".")!=10)
            {
                continue;
            }
            out<<timestamp<<" "<<src_ip<<" "<<dst_ip<<endl;
        }
        inf.close();
            // break;
    }
    cout<<"preprocess phase 1 done"<<endl;
    out.close();
    inf.open("processed/unprocessed_traffic.txt", ios::in);
    out.open("processed/processed_traffic.txt", ios::out);
    double start_time=-1;
    double end_time=-1;
    string timestamp,src_ip, dst_ip;
    deque<vector<string>> data;
    while (inf >> timestamp && inf >> src_ip && inf >> dst_ip) {
        if(start_time==-1)
        {
            start_time=stod(timestamp);
            end_time=stod(timestamp)+120;
        }
        if(stod(timestamp)>end_time)
        {
            for(int i=0;i<data.size();i++)
            {
                out<<data[i][0]<<" "<<data[i][1]<<" "<<data[i][2]<<endl;
            }
            end_time+=1;
            start_time+=1;
            while(!data.empty()&&stod(data.front()[0])<start_time)
            {
                data.pop_front();
            }
        }
        vector<string> temp;
        temp.push_back(timestamp);
        temp.push_back(src_ip);
        temp.push_back(dst_ip);
        data.push_back(temp);
    }
    cout<<"preprocess phase 2 done"<<endl;
    out.close();
    inf.close();
}